### PR TITLE
Fix NULL-class issues reported via static analysis

### PR DESCRIPTION
**Pull request by krcroft**
Not much to say; most involve simple checks. Interesting areas are where the code expects or branches differently based on a NULL pointer, such as in the shell commands (VER, PATH, CHOICE).

I've opted to split these commits into one-file-per-commit to simplify review and acceptance by up-stream. Although these could easily be squashed into one big commit.

Our new white-space pre-commit hook was also doing its job, with lots of trimming to these files.

### CODE REVIEW

**Comment by krcroft** [comment]
The null check in dr_flac is an internal fix.

Plan is to report this with a PR to upstream, and (assuming it's accepted), replace our dr_flac with the revised upstream version.

File: src/dos/dos_programs.cpp [hunk 1]
```diff
@@ -147,12 +148,12 @@ class MOUNT : public Program {
 
 			/* Change 8.3 to 11.0 */
 			char* dot = strchr(name,'.');
-			if(dot && (dot - name == 8) ) { 
+			if(dot && (dot - name == 8) ) {
 				name[8] = name[9];name[9] = name[10];name[10] = name[11];name[11] = 0;
 			}
```
Comment (dreamer) [inline src/dos/dos_programs.cpp]: (just a comment, but maybe there's a bug here) The more I look at this code, the more I think the original code in here was incorrect…

Normally, these lines should convert `abcdabcd.xyz` to `abcdabcdxyz`, right? But what about `abc.abcd.xyz`? This algorithm won't convert it, but perhaps it should be changed into `abc.abcdxyz`?

BTW, at least let's fix formatting of the whole if statement when touching this ;)

**Comment by krcroft** [comment]
Good point on the cleanup - I'll take a pass through all the trimmed lines and adjust formatting.

Interesting catch on that function.  I believe extra prior periods are (should) removed without replacement (which this function would miss like you caught).  `a.b.c` should become `ab.c`.  

This is from:

> If the filename contains characters not allowed in an 8.3 name (including space which was disallowed by convention though not by the APIs) or either part is too long, the name is stripped of invalid characters such as spaces and extra periods. If the name begins with periods . the leading periods are removed. Other characters such as + are changed to the underscore _, and letters are put in uppercase. The stripped name is then truncated to the first 6 letters of its basename, followed by a tilde, followed by a single digit, followed by a period ., followed by the first 3 characters of the extension.
Examples Example: TextFile.Mine.txt becomes TEXTFI~1.TXT (or TEXTFI~2.TXT, should TEXTFI~1.TXT already exist). ver+1.2.text becomes VER_12~1.TEX. .bashrc.swp becomes BASHRC~1.SWP

https://en.wikipedia.org/wiki/8.3_filename

File: src/dos/dos_programs.cpp (hunk 1 above)
Comment (krcroft) [reply src/dos/dos_programs.cpp]: Looks like this code prints the drive label, where **FAT volumes have the following limitations:**
- Contain as many as 11 character bytes and can include spaces, but no tabs.
- Volume labels cannot contain the following characters: ? / \ | . , ; : + = [ ] < > "
- Volume labels are stored as upper-case regardless of whether they contain lower-case letters.

Ref: https://en.wikipedia.org/wiki/Label_(command)

Comment (krcroft) [reply src/dos/dos_programs.cpp]: Addressed it as much as possible here: https://github.com/dreamer/dosbox-staging/commit/f9c8d719d043eaaba1127e3d4b2f8a1cd45895fd

This only performs a read-only cleanup of the label as-printed for the user when they run the 'mount' command.
The label itself is set in the 'Set_Label' function, which could be improved, but out-of-scope of this PR.

Here's a comparison of labels (as printed with `mount`) between SVN and this PR:

File: src/dos/cdrom_image.cpp [hunk 2]
```diff
@@ -803,10 +804,10 @@ bool CDROM_Interface_Image::LoadIsoFile(char* filename)
 	// try to detect iso type
 	if (CanReadPVD(track.file.get(), BYTES_PER_COOKED_REDBOOK_FRAME, false)) {
 		track.sectorSize = BYTES_PER_COOKED_REDBOOK_FRAME;
-		track.mode2 = false;
+		// track.mode2 = false (comment only, because this is the default value)
```
Comment (dreamer) [inline src/dos/cdrom_image.cpp]: Maybe replace with:
```
assert(track.mode2 == false);
```
?

File: src/hardware/mame/fmopl.cpp [hunk 3]
```diff
@@ -1958,23 +1958,23 @@ static void OPL_clock_changed(FM_OPL *OPL, uint32_t clock, uint32_t rate)
...
 {
-	char *ptr;
-	FM_OPL *OPL;
-	int state_size;
-
 	if (FM_OPL::LockTable(device) == -1) return 0;
 
 	/* calculate OPL state size */
-	state_size  = sizeof(FM_OPL);
+	int state_size  = sizeof(FM_OPL);
```
Comment (dreamer) [inline src/hardware/mame/fmopl.cpp]: I suggest `const int` :)

File: src/hardware/mame/ymf262.cpp [hunk 4]
```diff
@@ -2353,12 +2353,20 @@ static void OPL3ResetChip(OPL3 *chip)
 /* 'rate'  is sampling rate  */
 static OPL3 *OPL3Create(device_t *device, int clock, int rate, int type)
 {
+	// Guard
+	if (device == nullptr) {
+		return 0;
```
Comment (dreamer) [inline src/hardware/mame/ymf262.cpp]: `return nullptr`?

File: src/hardware/memory.cpp [hunk 5]
```diff
@@ -564,9 +564,6 @@ class MEMORY:public Module_base{
 			LOG_MSG("Stick with the default values unless you are absolutely certain.");
 		}
 		MemBase = new Bit8u[memsize*1024*1024];
-		if (!MemBase) E_Exit("Can't allocate main memory of %d MB",memsize);
```
Comment (dreamer) [inline src/hardware/memory.cpp]: (just a comment)
If we would like to preserve `E_Exit` with a helpful message in here, then line 566 could be rewritten as:

    MemBase = new (std::nothrow) uint8_t[memsize * 1024 * 1024];

File: src/misc/programs.cpp [hunk 6]
```diff
@@ -530,13 +539,22 @@ void CONFIG::Run(void) {
 			}
 			Section_line* sec = dynamic_cast <Section_line*>
 				(control->GetSection(std::string("autoexec")));
-
-			for(Bitu i = 0; i < pvars.size(); i++) sec->HandleInputline(pvars[i]);
+			if (!sec) {
+				WriteOut(MSG_Get("PROGRAM_CONFIG_SECTION_ERROR"));
+				return;
+			}
+			for(Bitu i = 0; i < pvars.size(); i++) {
```
Comment (dreamer) [inline src/misc/programs.cpp]: A space wants to squeak in between `for` and `(` :)

File: src/misc/support.cpp [hunk 7]
```diff
@@ -51,6 +52,15 @@ void trim(std::string &str) {
 	if (loc != std::string::npos) str.erase(loc+1);
 }
 
+void strip_punctuation(std::string &str) {
+	str.erase(
+		std::remove_if(
+			str.begin(),
+			str.end(),
+			[](unsigned char c){ return std::ispunct(c); }),
```
Comment (dreamer) [inline src/misc/support.cpp]: Even `unsigned char`, very nice :)

File: src/dos/dos_programs.cpp (hunk 1 above)
Comment (dreamer) [reply src/dos/dos_programs.cpp]: Looks so much better, thank you for thorough follow up and implementation :)

File: src/dos/cdrom_image.cpp (hunk 2 above)
Comment (krcroft) [reply src/dos/cdrom_image.cpp]: The perfect use of an assert! Got it.

File: src/hardware/mame/fmopl.cpp (hunk 3 above)
Comment (krcroft) [reply src/hardware/mame/fmopl.cpp]: There's an #if just after this that increments `state_size` 
(unless I'm missing something!)  leaving unresolved until confirmed.

File: src/hardware/mame/ymf262.cpp (hunk 4 above)
Comment (krcroft) [reply src/hardware/mame/ymf262.cpp]: Got it. Also added a couple other `nullptr` returns instead of `0` too.

File: src/hardware/memory.cpp (hunk 5 above)
Comment (krcroft) [reply src/hardware/memory.cpp]: Great suggestion: if we're going to hit allocation issues, it'll be here! 
Also added a couple other messages to adjacent (but surely smaller) allocations.

File: src/misc/programs.cpp (hunk 6 above)
Comment (krcroft) [reply src/misc/programs.cpp]: Got it, and did a general space-addition cleanup of all the `if (`, `for (`, and `while (` statements.  Although the inner-sections might not be perfect; I was getting dizzy after too much of it!

File: src/misc/support.cpp (hunk 7 above)
Comment (krcroft) [reply src/misc/support.cpp]: These lamdas are finally sinking in :-)

**Pull request merged**